dbt/include/athena/macros/materializations/models/table/table.sql

```diff
@@ -1,56 +1,121 @@
+-- TODO create a drop_relation_with_versions, to be sure to remove all historical versions of a table
 {% materialization table, adapter='athena' -%}
   {%- set identifier = model['alias'] -%}
 
   {%- set lf_tags = config.get('lf_tags', default=none) -%}
   {%- set lf_tags_columns = config.get('lf_tags_columns', default=none) -%}
   {%- set table_type = config.get('table_type', default='hive') | lower -%}
   {%- set old_relation = adapter.get_relation(database=database, schema=schema, identifier=identifier) -%}
+  {%- set is_ha = config.get('ha', default=false) -%}
+  {%- set s3_data_dir = config.get('s3_data_dir', default=target.s3_data_dir) -%}
+  {%- set s3_data_naming = config.get('s3_data_naming', default='table_unique') -%}
+  {%- set full_refresh_config = config.get('full_refresh', default=False) -%}
+  {%- set is_full_refresh_mode = (flags.FULL_REFRESH == True or full_refresh_config == True) -%}
+  {%- set external_location = config.get('external_location', default=none) -%}
   {%- set target_relation = api.Relation.create(identifier=identifier,
                                                 schema=schema,
                                                 database=database,
                                                 type='table') -%}
+  {%- set tmp_relation = api.Relation.create(identifier=target_relation.identifier ~ '__ha',
+                                             schema=schema,
+                                             database=database,
+                                             s3_path_table_part=target_relation.identifier,
+                                             type='table') -%}
+
+  {%- if (
+    table_type == 'hive'
+    and is_ha
+    and (s3_data_naming in ['table', 'table_schema'] or external_location is not none)
+  ) -%}
+      {%- set error_unique_location_hive_ha -%}
+          You need to have an unique table location when using table_hive_ha materialization.
+          Use s3_data_naming table_unique or schema_table_unique, and avoid to set an explicit
+          external_location.
+      {%- endset -%}
+      {% do exceptions.raise_compiler_error(error_unique_location_hive_ha) %}
+  {%- endif -%}
 
   {{ run_hooks(pre_hooks) }}
 
-  {%- if old_relation is none or table_type != 'iceberg' -%}
-    {%- if old_relation is not none -%}
-      {{ drop_relation(old_relation) }}
-    {%- endif -%}
+  {%- if table_type == 'hive' -%}
+
+    {%- if is_ha and not is_full_refresh_mode -%}
+      -- drop the tmp_relation
+      {% call statement('drop_tmp_relation', auto_begin=False) -%}
+        drop table if exists {{ tmp_relation.render_hive() }}
+      {%- endcall %}
+
+      -- create tmp table
+      {% call statement('main') -%}
+        {{ create_table_as(False, tmp_relation, sql) }}
+      {%- endcall %}
+
+      -- save the original target table location before swapping
+      {%- set original_table_location = adapter.s3_table_location(s3_data_dir,
```

Comment:
I believe that this can be removed, the original_table_location seems to be a left over on the original implementation.
Thanks to the `adapter.expire_glue_table_versions` data is expired based on the table location, so no need to keep this.